spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/AbstractMongoQuery.java

```diff
@@ -415,4 +429,55 @@ private Object deleteAndConvertResult(Query query, MongoEntityMetadata<?> metada
 			return writeResult != null ? writeResult.getN() : 0L;
 		}
 	}
+
+	/**
+	 * @author Thomas Darimont
+	 */
+	private class CursorBackedExecution extends Execution {
+
+		@Override
+		Object execute(final Query query) {
+
+			Class<?> javaType = getQueryMethod().getEntityInformation().getJavaType();
+			QueryMapper queryMapper = operations.getQueryMapper();
+
+			return createCursorAndWrapInCloseableIterator(query, javaType, queryMapper);
+		}
+
+		@SuppressWarnings("unchecked")
+		private CloseableIterator<Object> createCursorAndWrapInCloseableIterator(final Query query,
+				final Class<?> entityType, final QueryMapper queryMapper) {
+
+			final MongoPersistentEntity<?> persistentEntity = operations.getMappingContext().getPersistentEntity(entityType);
+
+			return (CloseableIterator<Object>) operations.execute(entityType, new CollectionCallback<Object>() {
```

Comment:
I am not sure the current setup here works as we expect it. Returning the adapter will basically just port the `DBCursor` to the client in adapter shape. Exceptions thrown while iterating over the result and mapping it still wouldn't be subject to the exception translation implemented in `MongoTemplate`.